### PR TITLE
[RFC] Add support for parser value_set to eBPF backend

### PR DESCRIPTION
**Pull request by tatry**
This PR adds the support for parser value_set to eBPF backend.

Co-authored-by: Tomasz Osiński <tomasz@opennetworking.org>
Co-authored-by: Mateusz Kossakowski <mateusz.kossakowski@orange.com>

### CODE REVIEW

File: backends/ebpf/psa/ebpfPsaParser.h [hunk 1]
```diff
@@ -30,13 +30,18 @@ class PsaStateTranslationVisitor : public StateTranslationVisitor {
  public:
     EBPFPsaParser * parser;
 
+    bool selectHasValueSet = false;
+    bool selectFirstIfStatement = true;
```
Comment (osinstom) [inline backends/ebpf/psa/ebpfPsaParser.h]: I suggest making these variables protected or private.

File: backends/ebpf/psa/ebpfPsaParser.cpp [hunk 2]
```diff
@@ -111,6 +193,154 @@ void PsaStateTranslationVisitor::compileVerify(const IR::MethodCallExpression *
     builder->blockEnd(true);
 }
 
+// =====================EBPFValueSetPSA=============================
+EBPFValueSetPSA::EBPFValueSetPSA(const EBPFProgram* program, const IR::P4ValueSet* p4vs,
+                                 cstring instanceName, CodeGenInspector* codeGen)
+        : EBPFTableBase(program, instanceName, codeGen), size(0), pvs(p4vs) {
+    CHECK_NULL(pvs);
+    valueTypeName = "u32";  // value is not used, we will check only if entry exists
```
Comment (kmateuszssak) [inline backends/ebpf/psa/ebpfPsaParser.cpp]: Is this possible to not use `valueTypeName` to check if entry exists?

File: backends/ebpf/psa/README.md [hunk 3]
```diff
@@ -270,6 +270,18 @@ If a deparser triggers the `pack()` method, an eBPF program inserts data defined
 A user space application is responsible for performing periodic queries to this map to read a Digest message. It can use either
 `psabpf-ctl digest get pipe`, `psabpf_digest_get_next` from psabpf C API or `bpf_map_lookup_and_delete_elem` from `libbpf` API.
 
+### value_set
+
+[value_set](https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-value-set) is a P4 lang construct allowing to determine next
+parser state based on runtime values. To achieve this compiler generates additional hash map. In select case expression
```
Comment (osinstom) [inline backends/ebpf/psa/README.md]:
```suggestion
parser state based on runtime values. The P4-eBPF compiler generates additional hash map for each `ValueSet` instance. In select case expression
```

File: backends/ebpf/psa/README.md [hunk 4]
```diff
@@ -270,6 +270,18 @@ If a deparser triggers the `pack()` method, an eBPF program inserts data defined
...
+it is checked whether entry for given key exists in this map. Value of this map is ignored and not used.
```
Comment (osinstom) [inline backends/ebpf/psa/README.md]:
```suggestion
Each `select()` on `ValueSet` is translated into a lookup into the BPF hash map to check if an entry for a given key exists. A value of the BPF map is ignored. 
```

File: backends/ebpf/psa/README.md [hunk 5]
```diff
@@ -270,6 +270,18 @@ If a deparser triggers the `pack()` method, an eBPF program inserts data defined
 A user space application is responsible for performing periodic queries to this map to read a Digest message. It can use either
 `psabpf-ctl digest get pipe`, `psabpf_digest_get_next` from psabpf C API or `bpf_map_lookup_and_delete_elem` from `libbpf` API.
 
+### value_set
+
+[value_set](https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-value-set) is a P4 lang construct allowing to determine next
+parser state based on runtime values. To achieve this compiler generates additional hash map. In select case expression
+it is checked whether entry for given key exists in this map. Value of this map is ignored and not used. 
+
+Select expression which is referred by instance of `value_set`, have to be replaced by chain of `if ... else ...` statements
```
Comment (osinstom) [inline backends/ebpf/psa/README.md]: this paragraph is too detailed IMO, can be removed

File: backends/ebpf/psa/README.md [hunk 6]
```diff
@@ -378,6 +390,7 @@ with some NICs. So far, we have verified the correct behavior with Intel 82599ES
 - `psa_idle_timeout` is not supported yet.
 - DirectCounter and DirectMeter externs are not supported for P4 tables with implementation (ActionProfile).
 - The `xdp2tc=head` mode works only for packets larger than 34 bytes (the size of Ethernet and IPv4 header).
+- `value_set` supports only exact match type and can match only single field in select expression.
```
Comment (osinstom) [inline backends/ebpf/psa/README.md]:
```suggestion
- `value_set` only supports the exact match type and can only match on a single field in the `select()` expression.
```

File: backends/ebpf/psa/README.md [hunk 7]
```diff
@@ -390,7 +403,7 @@ The TC subsystem enables implementation of the full PSA specification, contrary
 contribute the XDP-based version of P4-eBPF that is not fully spec-compliant, but provides higher throughput.
 - **XDP2TC mode.** the generated eBPF programs use `bpf_xdp_adjust_meta()` to transfer original EtherType from XDP to TC. This mode 
 may not be supported by some NIC drivers. We will add two other modes that can be used alternatively. 
-- **ValueSet support.** The parser generated by P4-eBPF does not support ValueSet. We plan to contribute ValueSet implementation for eBPF.
+- **ValueSet support.** We plan to extend implementation to support other match kinds and more fields in select expression.
```
Comment (osinstom) [inline backends/ebpf/psa/README.md]:
```suggestion
- **ValueSet support.** We plan to extend implementation to support other match kinds and multiple fields in the `select()` expression.
```

**Pull request closed without merging**